### PR TITLE
Toggle between admin and judge role

### PR DESCRIPTION
**Pull request by kayladavis**
In this PR I added actions/reducers to toggle judge/admin roles and produce the admin or judge view based on the active performing role.

I also realized a bug with the way I wrote voting (admins need to vote too). So that's fixed an tested.

### CODE REVIEW

File: frontend/src/Admin/components/NavBar.js [hunk 1]
```diff
@@ -35,6 +38,9 @@ export default class NavBar extends Component {
           <NavbarToggler onClick={this.toggle} />
           <Collapse isOpen={this.state.isOpen} navbar>
             <Nav className="ml-auto" navbar>
+              <NavItem>
+                <NavLink tag={Link} to='/' onClick={this.props.switchToJudge}>Switch to: Judge</NavLink>
```
Comment (chtinahow) [inline frontend/src/Admin/components/NavBar.js]: Quick thing,I think that the Switch to: Judge looks weird compared to the rest of the links on the navbar. No big deal if others disagree

File: frontend/src/Admin/components/NavBar.js [hunk 2]
```diff
@@ -51,3 +57,13 @@ export default class NavBar extends Component {
     )
   }
 }
+
+
+const mapDispatchToProps = (dispatch) => ({
+  switchToJudge: () => dispatch(switchToJudge())
```
Comment (robmcl4) [inline frontend/src/Admin/components/NavBar.js]: Can you stick this instead in [`frontend/src/Admin/containers/NavBar.js`](https://github.com/robmcl4/A-B-S-T-R-A-C-T-I-O-N/blob/master/frontend/src/Admin/containers/NavBar.js#L8), we've been leaving the "components" folder free of redux hookup

File: frontend/src/Judge/components/NavBar.js [hunk 3]
```diff
@@ -45,3 +51,12 @@ export default class NavBar extends Component {
     )
   }
 }
+
+const mapDispatchToProps = (dispatch) => ({
+  switchToAdmin: () => dispatch(switchToAdmin())
```
Comment (robmcl4) [inline frontend/src/Judge/components/NavBar.js]: same as before, can you stick the redux hookup in the "containers" folder

File: frontend/src/shared/reducers.js [hunk 4]
```diff
@@ -33,6 +33,21 @@ const auth = (state = {}, action) => {
...
+      return newState  
+    default:
+      return state
+  }
+}
+
 export default combineReducers({
-  auth
+  auth,
+  performingRole
```
Comment (robmcl4) [inline frontend/src/shared/reducers.js]: this smells a bit odd to me... I feel like `performingRole` should be a a key in the `auth` reducer that maps to a string, such that auth looks like:

```json
{
   "token": "...",
   "user": {
     "username": "souperManne",
     "firstName": "Clark",
     "lastName": "Kent",
     "type": "ADMIN",
     "createdAt": "...",
     "iat": 1516390788
  },
  "performingRole": "JUDGE"
}
```

... which I think might be a bit cleaner for everything, especially considering the state currently doesn't clear when you logout (bug or feature tho?)

File: frontend/src/shared/components/LoginSwitch.js [hunk 5]
```diff
@@ -56,12 +63,14 @@ class LoginSwitch extends Component {
...
-  user: state.shared.auth.user
+  user: state.shared.auth.user,
+  performingRole: state.shared.performingRole
 })
 
 const mapDispatchToProps = (dispatch) => ({
   shouldLogin,
-  login: () => dispatch(login())
+  login: () => dispatch(login()),
+
```
Comment (robmcl4) [inline frontend/src/shared/components/LoginSwitch.js]: nit: extra line

File: frontend/src/shared/reducers.js [hunk 6]
```diff
@@ -33,6 +33,21 @@ const auth = (state = {}, action) => {
   }
 }
 
+const performingRole = (state = {type: 'ADMIN'}, action) => {
+  let newState = state
+  switch (action.type) {
+    case actions.JUDGE_ROLE:
+      newState.type = 'JUDGE'
+      return newState
```
Comment (robmcl4) [inline frontend/src/shared/reducers.js]:
```javascript
return {type: 'JUDGE'}
```

File: frontend/src/shared/reducers.js (hunk 4 above)
Comment (kayladavis) [reply frontend/src/shared/reducers.js]: I think moving it to auth could work. This task was mostly me learning to use redux and I just wasn't thinking of designing it that way when I started.

File: frontend/src/Admin/components/NavBar.js [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import React, { Component } from 'react'
 import PropTypes from 'prop-types'
 import { Link } from 'react-router-dom'
+import { connect } from 'react-redux'
```
Comment (billdybas) [inline frontend/src/Admin/components/NavBar.js]: You can remove this, now.

File: frontend/src/Admin/components/NavBar.js [hunk 8]
```diff
@@ -10,14 +11,15 @@ const NavBarContainer = styled.div`
...
+    logout: PropTypes.func.isRequired,
+    switchToJudge: PropTypes.func.isRequired,
   }
 
   constructor (props) {
     super(props)
 
     this.state = {
-      isOpen: false
+      isOpen: false,
```
Comment (billdybas) [inline frontend/src/Admin/components/NavBar.js]: nit: extra comma

File: frontend/src/Judge/components/NavBar.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import React, { Component } from 'react'
 import PropTypes from 'prop-types'
 import { Link } from 'react-router-dom'
+import { connect } from 'react-redux'
```
Comment (billdybas) [inline frontend/src/Judge/components/NavBar.js]: You can remove this, now.

File: frontend/src/shared/components/LoginSwitch.js [hunk 10]
```diff
@@ -33,14 +35,20 @@ class LoginSwitch extends Component {
     const {
       user
     } = this.props
+    const performingRole = this.props.performingRole
```
Comment (billdybas) [inline frontend/src/shared/components/LoginSwitch.js]: You can destructure this like the line above it:

```js
const {
  user,
  performingRole
} = this.props
```

File: frontend/src/shared/components/LoginSwitch.js [hunk 11]
```diff
@@ -56,12 +64,13 @@ class LoginSwitch extends Component {
...
 const mapStateToProps = (state) => ({
-  user: state.shared.auth.user
+  user: state.shared.auth.user,
+  performingRole: state.shared.auth.performingRole
 })
 
 const mapDispatchToProps = (dispatch) => ({
   shouldLogin,
-  login: () => dispatch(login())
+  login: () => dispatch(login()),
```
Comment (billdybas) [inline frontend/src/shared/components/LoginSwitch.js]: nit: extra comma

File: frontend/src/shared/actions.js [hunk 12]
```diff
@@ -3,6 +3,8 @@ import { push } from 'connected-react-router'
 
 export const LOGIN_USER = 'LOGIN_USER'
 export const LOGOUT_USER = 'LOGOUT_USER'
+export const JUDGE_ROLE = 'JUDGE_ROLE'
+export const ADMIN_ROLE = 'ADMIN_ROLE'
```
Comment (billdybas) [inline frontend/src/shared/actions.js]: I think I would prefer these actions to be named:

```js
export const SWITCH_TO_JUDGE = 'SWITCH_TO_JUDGE'
export const SWITCH_TO_ADMIN = 'SWITCH_TO_ADMIN'
```

Don't forget to update the references in other files.

File: frontend/src/Admin/components/NavBar.js (hunk 1 above)
Comment (billdybas) [reply frontend/src/Admin/components/NavBar.js]: I think I would rename it to "View as Judge"

File: frontend/src/Judge/components/NavBar.js [hunk 13]
```diff
@@ -35,6 +37,9 @@ export default class NavBar extends Component {
           <NavbarToggler onClick={this.toggle} />
           <Collapse isOpen={this.state.isOpen} navbar>
             <Nav className="ml-auto" navbar>
+              <NavItem>
+                <NavLink tag={Link} to='/' onClick={this.props.switchToAdmin}>Switch to: Admin</NavLink>
```
Comment (billdybas) [inline frontend/src/Judge/components/NavBar.js]: I would rename this to "View as Admin"

Comment (billdybas) [reply frontend/src/Judge/components/NavBar.js]: Also, I would add some sort of visual separation between this and the other links, say a `|` symbol in between them.

**Pull request merged**